c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -482,19 +484,131 @@ class CachingAllocatorConfig {
+          planned_ptr); // find the closest start/end point of a previous
+                        // allocation after the current allocate ptr
+      // perform the allocation only if it fails inside the block assigned for
+      // plan execution and it does not overlap with previuosly allocated events
+      // if it_ptr_lower_bound points to an end point, this means the current
+      // allocation starts in the middle of a previous allocation; do not
+      // allocate. If it_ptr_lower_bound points to a start point, check if
+      // planned_ptr + size <= start point to ensure no overlap
+      if ((planned_ptr + size <= planning_size[device]) &&
+          ((it_ptr_lower_bound == allocated_blocks[device].end()) ||
+           (it_ptr_lower_bound->second &&
+            (planned_ptr + size) <= (uint32_t)it_ptr_lower_bound->first))) {
+        block = new Block(
+            device,
+            stream,
+            size,
+            nullptr,
+            reinterpret_cast<void*>(
+                reinterpret_cast<intptr_t>(plan_blk_ptr[device]->ptr) +
+                planned_ptr)); // allocation block ptr is planning block ptr +
+                               // allocation offset according to plan
+        block->planned = true; // marks the block as allocated by plan
+        allocated_blocks[device][planned_ptr] =
+            true; // store start point of allocation
+        allocated_blocks[device][planned_ptr + size - 1] =
+            false; // store end point of allocation
+      }
+      plan_counter[device] += 1; // wait for next allocation
+      // if we reach end of plan, restart by ignoring non-recurring events
+      if (plan_counter[device] == mem_plan[device].size()) {
+        plan_counter[device] = non_rec_length[device] + 1;
+      }
+    }
+    return block; // nullptr means allocation is not performed by plan, hence,
+                  // use CudaCachingAllocator
+  }
+  void free(Block* block) {
+    // remove start and end points of the event from tree
+    allocated_blocks[block->device].erase(
+        reinterpret_cast<intptr_t>(block->ptr) -
+        reinterpret_cast<intptr_t>(plan_blk_ptr[block->device]->ptr));
+    allocated_blocks[block->device].erase(
+        reinterpret_cast<intptr_t>(block->ptr) + block->size -
+        reinterpret_cast<intptr_t>(plan_blk_ptr[block->device]->ptr) - 1);
+    delete block;
+  }
+};
+MemoryPlanner memory_planner;
+
 // records allocate/free events
 class MemoryEventTracker {
  private:
   mutable std::mutex mutex;
 
  public:
   std::vector<std::vector<AllocFreeEvent>> alloc_free_events;
+
+  bool active = false;
   void append_alloc_free_event(intptr_t ptr, int size, int device) {
     std::lock_guard<std::mutex> lock(mutex);
-    alloc_free_events[device].push_back(AllocFreeEvent{
-        ptr, // ptr
-        size, // size: of allocation in bytes
-    });
+    // if event is created by initialization or raw_alloc, do not create a new
```

Comment:
which initialization are you referring? you mean the `resize(1)` below?

Reply:
Yes, the resize(1). I changed it.